File: output/cpp/tests/actions_call_wrappers.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

class CallMismatchFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    CallMismatchFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class CallMismatchFrameContext {
public:
    CallMismatchFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    CallMismatchFrameContext(CallMismatchFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class CallMismatchCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<CallMismatchFrameEvent> forward_event;
    std::unique_ptr<CallMismatchCompartment> parent_compartment;

    explicit CallMismatchCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<CallMismatchCompartment> clone() const {
        auto c = std::make_unique<CallMismatchCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class CallMismatch {
private:
    std::vector<std::unique_ptr<CallMismatchCompartment>> _state_stack;
    std::unique_ptr<CallMismatchCompartment> __compartment;
    std::unique_ptr<CallMismatchCompartment> __next_compartment;
    std::vector<CallMismatchFrameContext> _context_stack;

    void __kernel(CallMismatchFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            CallMismatchFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                CallMismatchFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    CallMismatchFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(CallMismatchFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "S") {
            _state_S(__e);
        }
    }

    void __transition(std::unique_ptr<CallMismatchCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_S(CallMismatchFrameEvent& __e) {
        if (__e._message == "e") {
            this->handle();
        }
    }

    void log(std::string message) {
                    // log sink
                    this->last = message;
    }

    void handle() {
                    // Calls 'log' without _action_ prefix; wrappers should preserve FRM names
                    this->log(std::string("hello"));
    }

public:
    std::string last = "";

    CallMismatch() {
        __compartment = std::make_unique<CallMismatchCompartment>("S");
        CallMismatchFrameEvent __frame_event("$>");
        CallMismatchFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void e() {
        CallMismatchFrameEvent __e("e");
        CallMismatchFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

// Stub functions for placeholder calls
void native() {}
void x() {}

// TAP test harness
int main() {
    printf("TAP version 14\n");
    printf("1..1\n");
    try {
        CallMismatch s;
        s.e();
        printf("ok 1 - actions_call_wrappers\n");
    } catch (...) {
        printf("not ok 1 - actions_call_wrappers\n");
    }
    return 0;
}

File: output/java/tests/actions_call_wrappers.java
import java.util.*;


import java.util.*;

class CallMismatchFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    CallMismatchFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    CallMismatchFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class CallMismatchFrameContext {
    CallMismatchFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    CallMismatchFrameContext(CallMismatchFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class CallMismatchCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    CallMismatchFrameEvent forward_event;
    CallMismatchCompartment parent_compartment;

    CallMismatchCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    CallMismatchCompartment copy() {
        CallMismatchCompartment c = new CallMismatchCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class CallMismatch {
    private ArrayList<CallMismatchCompartment> _state_stack;
    private CallMismatchCompartment __compartment;
    private CallMismatchCompartment __next_compartment;
    private ArrayList<CallMismatchFrameContext> _context_stack;
    public String last = "";

    public CallMismatch() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new CallMismatchCompartment("S");
        __next_compartment = null;
        CallMismatchFrameEvent __frame_event = new CallMismatchFrameEvent("$>");
        CallMismatchFrameContext __ctx = new CallMismatchFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(CallMismatchFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            CallMismatchFrameEvent exit_event = new CallMismatchFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                CallMismatchFrameEvent enter_event = new CallMismatchFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    CallMismatchFrameEvent enter_event = new CallMismatchFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(CallMismatchFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("S")) {
            _state_S(__e);
        }
    }

    private void __transition(CallMismatchCompartment next) {
        __next_compartment = next;
    }

    public void e() {
        CallMismatchFrameEvent __e = new CallMismatchFrameEvent("e");
        CallMismatchFrameContext __ctx = new CallMismatchFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void _state_S(CallMismatchFrameEvent __e) {
        if (__e._message.equals("e")) {
            this.handle();
        }
    }

    private void log(Object message) {
                    // log sink
                    this.last = message;
    }

    private void handle() {
                    // Calls 'log' without _action_ prefix; wrappers should preserve FRM names
                    this.log("hello");
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("TAP version 14");
        System.out.println("1..1");
        try {
            CallMismatch s = new CallMismatch();
            s.e();
            System.out.println("ok 1 - actions_call_wrappers");
        } catch (Exception ex) {
            System.out.println("not ok 1 - actions_call_wrappers # " + ex);
        }
    }
}
